lib/index.ts

```diff
@@ -526,63 +531,70 @@ export class Worker {
 			onMessageEventHandler,
 		);
 
-		const [defaultTypeContracts, nonTypeContracts] = _.partition(
-			Object.values(contracts),
-			(contract) => {
-				return contract.type.split('@')[0] === 'type';
-			},
-		);
+		await this.kernel.runMigration(
+			logContext,
+			packageName,
+			packageVersion,
+			async (_context) => {
+				const [defaultTypeContracts, nonTypeContracts] = _.partition(
+					Object.values(contracts),
+					(contract) => {
+						return contract.type.split('@')[0] === 'type';
+					},
+				);
 
-		const [defaultLoopContracts, defaultContracts] = _.partition(
-			Object.values(nonTypeContracts),
-			(contract) => {
-				return contract.type.split('@')[0] === 'loop';
-			},
-		);
+				const [defaultLoopContracts, defaultContracts] = _.partition(
+					Object.values(nonTypeContracts),
+					(contract) => {
+						return contract.type.split('@')[0] === 'loop';
+					},
+				);
 
-		// Insert type contracts as prerequisite
-		await Promise.all(defaultTypeContracts.map(checkThenReplaceContract));
+				// Insert type contracts as prerequisite
+				await Promise.all(defaultTypeContracts.map(checkThenReplaceContract));
 
-		// Insert loop contracts as prerequisite
-		await Promise.all(defaultLoopContracts.map(checkThenReplaceContract));
+				// Insert loop contracts as prerequisite
+				await Promise.all(defaultLoopContracts.map(checkThenReplaceContract));
 
-		// Insert other worker contracts
-		await Promise.all(defaultContracts.map(checkThenReplaceContract));
+				// Insert other worker contracts
+				await Promise.all(defaultContracts.map(checkThenReplaceContract));
 
-		await Promise.all(
-			actions.map((action) => {
-				return checkThenReplaceContract(action.contract);
-			}),
-		);
+				await Promise.all(
+					actions.map((action) => {
+						return checkThenReplaceContract(action.contract);
+					}),
+				);
 
-		// Get all contracts provided by plugins
-		const pluginContracts = this.pluginManager.getCards();
+				// Get all contracts provided by plugins
```

Comment:
Shouldn't upsert plugin contracts in this callback as this callback is tied to the worker version and not plugin library versions. Might just move this out for now, or execute `runMigration()` for each plugin and its version.

Reply:
Hmm, I may need to add an option to `kernel.runMigrations()` to specify that the callback should be executed regardless of the `version` value in the `migrations` table (for `name=@balena/jellyfish-worker` in this case). I'm not sure how to cleanly handle all of the following cases on worker init:
- new worker version, no new plugin version(s)
- new version of worker, new plugin version(s)
- no new version of worker, new plugin version(s)

The first two are simple enough as the current logic is tied to the worker version. It's a bit difficult as the contracts provided by plugins depend on those provided by the worker. I'll sleep on it and maybe have a better solution in the morning, but having a single instance unconditionally execute all of these contract checks + upserts is a lot better than having all worker instances unconditionally execute them.

Reply:
I'm now thinking adding an optional parameter to `kernel.runMigration()` that defines a list of dependency name(s) and version(s) that must be semver satisified in the `migrations` table before the migration callback is executed. Which means, we might could then execute `kernel.runMigration()` for each plugin and specify `@balena/jellyfish-worker` and its current version as a dependency.

Reply:
Is the `requires` parameter on `runMigration` really needed?

Maybe I don't see what would be the problem with the approach in the current PR

I like the idea of updating all plugins independently, looks simpler.

When I had to create the "relationships" contracts I ordered the plugins by their dependencies. Maybe we could apply this idea here? I need to look for the exact "dependency tree", but it was IIRC worker -> default-plugin -> the rest

Reply:
And, if the `requires` is not met, shouldn't the migration be aborted? As the new version of the code will run against an old version of the schema

Reply:
> Is the requires parameter on runMigration really needed?

The worker currently upserts all contracts provided by the plugins, so if we don't have the ability to specify that plugins require the current version of the workers migration record to be up to date we can miss plugin migration executions if only the plugin was updated and not the worker as well.

> When I had to create the "relationships" contracts I ordered the plugins by their dependencies.
We still technically support dependencies amongst plugins, but don't have any at the moment as `plugin-default` and `plugin-product-os` were merged into the worker. This code should also therefore support plugin dependencies when defining the `requires` passed to `kernel.runMigration()`. I'll look at building `requires`  as well as deciding migration execution order based on defined plugin dependencies.

> And, if the requires is not met, shouldn't the migration be aborted? As the new version of the code will run against an old version of the schema
This is done on the autumndb side of things. If `requires` isn't satisfied, the migration is canceled and the lock on the migrations table is dropped.

Reply:
> When I had to create the "relationships" contracts I ordered the plugins by their dependencies. 
Is this in code somewhere? I looked but couldn't find it. If it is in code, maybe we can reuse it here.

Reply:
> I'll look at building requires as well as deciding migration execution order based on defined plugin dependencies.

Added `pluginManager.getPlugins()` function that returns loaded plugins in order of their dependency (`requires`) needs. Also added a unit test for it.